include/revng/ADT/GenericGraph.h

```diff
@@ -695,13 +695,13 @@ class MutableEdgeNode
 
 public:
   EdgeView addSuccessor(MutableEdgeNode *NewSuccessor, EdgeLabel EL = {}) {
-    auto [Owner, View] = constructEdge(this, NewSuccessor, std::move(EL));
+    auto &&[Owner, View] = constructEdge(this, NewSuccessor, std::move(EL));
```

Comment:
I'm not so sure about all these changes. We might be moving things that previously were copied.

I'd be much more in favor of adding `const auto &` wherever possible, but this seems risky.

Reply:
I'll take another look!

Reply:
Wait, I think there's a misunderstanding. `auto &&` is not a _rvalue_ reference, it's a _universal_ reference.

Here's a bit of an illustration: https://godbolt.org/z/Ghd7qh6ca

TL;DR is that because of decay, moving will never occur without an explicit `std::move`, but it still lets the compiler avoid unnecessary moves (compare `by-value` to `by-rvalue-reference`).

Reply:
> auto && is not a rvalue reference, it's a universal reference

Ah OK.
But I get the exact same results without `&&`... Can you make an example where using `auto &&` is beneficial over just `auto`?

Reply:
Yep, that's how how this entire conversation started: I thought `auto []` was always sufficient, but that was not the case:

https://github.com/revng/revng/pull/412#discussion_r1858212292